src/types.jl

```diff
@@ -57,6 +57,10 @@ struct BoundaryConstraint{variable_dependence}
     f::Function
 end
 
+struct BoundaryConstraint!{variable_dependence}
+    f!::Function
+end
```

Comment:
@jbcaillau `f!::Function` will not be efficient because `Function` is a generic type.
You need a parametric structure like this:
```julia
struct BoundaryConstraint!{F}
    f!::F
end
```

Reply:
> @jbcaillau `f!::Function` will not be efficient because `Function` is a generic type. You need a parametric structure like this:
> 
> ```julia
> struct BoundaryConstraint!{F}
>     f!::F
> end
> ```

@amontoison Instead of 

```julia
struct CallableType
    f::Function
end

(F::CallableType)(n::Int) = F.f(n)

F = CallableType(n->2n)
```

Do you recommend this?

```julia
struct CallableType{T}
    f::T
end

(F::CallableType)(n::Int) = F.f(n)

f = n->2n
F = CallableType{typeof(f)}(f)
```

Reply:
Yes, and you can avoid to specify `typeof(f)`, Julia will do that automatically for you:
```julia
struct CallableType{T}
    f::T
end

(F::CallableType)(n::Int) = F.f(n)

f = n->2n
F = CallableType(f)
```
By having a parametric structure `CallableType`, Julia doesn't need to re-infer the type of `F.f` at every call. After the first call, a compiled version is available, and future calls to `CallableType` are faster.

Reply:
And if we have to types. How does it work?

```julia
struct CallableType{T, U}
    f::T
end

f = n->2n
F = CallableType{SomeType}(f)
```

Is this better ?

```julia

Reply:
What is the goal of `U`? It is for dispatch like a symbol `:inplace` `:not_inplace`?
If you don't have any attribute of type `U`, it should not be used in practice.
The second is the best one
```
struct CallableType{T <: Function}
    f::T
end
```
But I'm wondering if you just need a "functor": https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects

Reply:
> What is the goal of `U`? It is for dispatch like a symbol `:inplace` `:not_inplace`? If you don't have any attribute of type `U`, it should not be used in practice. The second structure is the best one
> 
> ```julia
> struct CallableType{T <: Function}
>     f::T
> end
> ```
> 
> But I'm wondering if what you want is not just a "functor": https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects

Yes exactly, the role of `U` is for dispatch. For instance, in-place vs out-of-place, autonomous vs non-automous, etc.

Reply:
Ok, I understand what you are doing, so what you need is:
```
 struct BoundaryConstraint{variable_dependence, T <: Function} 
     f::T
 end
```

Reply:
You want to keep `variable_dependence` as the first parametric type because you do the dispatch on it and not the other one.

Reply:
@amontoison thanks for the tip. about the motivation: allow different signatures from the user (depending on whether there is a variable or not, *etc.*) and generate internally the most general signature that can be used uniformly afterwards.

Reply:
> You want to keep `variable_dependence` as the first parametric type because you do the dispatch on it and not the other one.

So dispatch processes the arguments starting from the first until it can determine a match ?

Reply:
Yes, and if you are not doing anything special for the last arguments, you can avoid parametrizing your functions with all arguments and only parametrize with the first one. This will avoid compiling a method for each combination if it's not needed. I think it's explained somewhere in Julia's documentation (over-parametrization).

Reply:
@amontoison a simple example/illustration somewhere?

Reply:
https://gensoft.pasteur.fr/docs/julia/1.4.1/manual/performance-tips.html#The-dangers-of-abusing-multiple-dispatch-(aka,-more-on-types-with-values-as-parameters)-1

Reply:
Catching up on this today, this is exactly what we discussed yesterday @ocots